subprojects/core/src/main/java/org/gradle/api/internal/credentials/GradlePropertiesCredentialsProviderFactory.java

```diff
@@ -0,0 +1,200 @@
+            }
+            return propertyProvider.getOrNull();
+        }
+
+        @Nullable
+        String getOptionalProperty(String property) {
+            String identityProperty = identityProperty(property);
+            return providerFactory.gradleProperty(identityProperty).forUseAtConfigurationTime().getOrNull();
+        }
+
+        void assertRequiredValuesPresent() {
+            if (!missingProperties.isEmpty()) {
+                TreeFormatter errorBuilder = new TreeFormatter();
+                errorBuilder.node("The following Gradle properties are missing for '").append(identity).append("' credentials");
+                errorBuilder.startChildren();
+                for (String missingProperty : missingProperties) {
+                    errorBuilder.node(missingProperty);
+                }
+                errorBuilder.endChildren();
+                missingProperties.clear();
+                throw new MissingValueException(errorBuilder.toString());
+            }
+        }
+
+        private String identityProperty(String property) {
+            return identity + property;
+        }
+    }
+
+    private class PasswordCredentialsProvider extends CredentialsProvider<PasswordCredentials> {
+
+        PasswordCredentialsProvider(String identity) {
+            super(identity);
+        }
+
+        @Override
+        public PasswordCredentials call() {
+            String username = getRequiredProperty("Username");
+            String password = getRequiredProperty("Password");
+            assertRequiredValuesPresent();
+
+            return new DefaultPasswordCredentials(username, password);
+        }
+    }
+
+    private class AwsCredentialsProvider extends CredentialsProvider<AwsCredentials> {
+
+        AwsCredentialsProvider(String identity) {
+            super(identity);
+        }
+
+        @Override
+        public AwsCredentials call() {
+            String accessKey = getRequiredProperty("AccessKey");
+            String secretKey = getRequiredProperty("SecretKey");
+            assertRequiredValuesPresent();
+
+            AwsCredentials credentials = new DefaultAwsCredentials();
+            credentials.setAccessKey(accessKey);
+            credentials.setSecretKey(secretKey);
+            credentials.setSessionToken(getOptionalProperty("SessionToken"));
+            return credentials;
+        }
+    }
+
+    private static void validateIdentity(@Nullable String identity) {
+        if (identity == null || identity.isEmpty() || !identity.chars().allMatch(Character::isLetterOrDigit)) {
+            throw new IllegalArgumentException("Identity may contain only letters and digits, received: " + identity);
+        }
+    }
+
+    private <T extends Credentials> Provider<T> evaluateAtConfigurationTime(Callable<T> provider) {
+        return new InterceptingProvider<>(provider);
+    }
+
+    private class InterceptingProvider<T> extends DefaultProvider<T> {
```

Comment:
We generally should avoid adding more implementations of `Provider`. Why did `DefaultProvider` not work here?

Reply:
I extended `DefaultProvider` here in order to intercept when the provider gets its value evaluated. This is used to catch the evaluations that happen during configuration time which indicate that the value of this provider will be needed during task execution somewhere down the line. 
Any missing values are noted here and a task execution listener uses this data to fail fast before execution starts if we know that some of the credentials will be needed and are missing.
This implementation is private for this specific use case. I could not figure a better way.

Reply:
I think we want to avoid anyone querying the credentials until they are needed at execution time, rather than relying on them being queried eagerly during configuration.

I would infer that a particular set of credentials are required by a given task (and other things like build services later) when we traverse the `@Input` properties while calculating the task graph, which is basically when `getProducer()` is called. This means we will also know which tasks depend on a given set of credentials, which allows other interesting use cases, such having a task produce the credentials rather than using our convention, or starting to run other tasks while prompting the user for particular credentials.